apps-rendering/src/server/insertNewsletter.ts

```diff
@@ -0,0 +1,226 @@
+
+type TextElementDescription = {
+	isError: false;
+	kind: ElementKind.Text;
+	paragraphs: number;
+	text: Text;
+};
+type OtherElementDescription = {
+	isError: false;
+	paragraphs: 0;
+	kind: Exclude<ElementKind, ElementKind.Text>;
+};
+type ErrorDescription = {
+	isError: true;
+	paragraphs: 0;
+};
+
+type BodyElementDescription =
+	| TextElementDescription
+	| OtherElementDescription
+	| ErrorDescription;
+
+const wrapInResult = (element: BodyElement): Result<string, BodyElement> => ({
+	kind: ResultKind.Ok,
+	value: element,
+});
+
+const getBodyDescription = (
+	body: Array<Result<string, BodyElement>>,
+): BodyElementDescription[] => {
+	return body.map((_) => {
+		if (_.kind === ResultKind.Err) {
+			return {
+				isError: true,
+				paragraphs: 0,
+			};
+		}
+
+		if (_.value.kind === ElementKind.Text) {
+			const paragraphs = _.value.doc.childElementCount;
+			return {
+				isError: false,
+				kind: _.value.kind,
+				paragraphs,
+				text: _.value,
+			};
+		}
+
+		return {
+			isError: false,
+			kind: _.value.kind,
+			paragraphs: 0,
+		};
+	});
+};
+
+const findWhereToInsert = (
+	body: Array<Result<string, BodyElement>>,
+	fraction = 1 / 2,
+): {
+	bodyResultIndex: number;
+	bodyResultParagraphIndex: number;
+	text: Text;
+} | null => {
+	const bodyDescription = getBodyDescription(body);
+	const totalParagraphs = bodyDescription.reduce(
+		(previous: number, current) => previous + current.paragraphs,
+		0,
+	);
+	const targetPlace = Math.floor(totalParagraphs * fraction);
+
+	let paragraphsFromPreviousElements = 0;
+	for (let i = 0; i < bodyDescription.length; i++) {
+		const current = bodyDescription[i];
+
+		if (current.isError || current.kind !== ElementKind.Text) {
+			continue;
+		}
+
+		if (paragraphsFromPreviousElements + current.paragraphs < targetPlace) {
+			paragraphsFromPreviousElements += current.paragraphs;
+			continue;
+		}
+
+		return {
+			bodyResultIndex: i,
+			bodyResultParagraphIndex:
+				targetPlace - paragraphsFromPreviousElements,
+			text: current.text,
+		};
+	}
+
+	return null;
+};
+
+/**
+ * Splits a Text BodyElement into to parts after a give number of
+ * nodes in the html and puts another BodyElement between them.
+ *
+ * @param textElement a Text BodyElement from the original Item Body
+ * @param elementToInsert A Body element to place inside the text
+ * @param splitPoint the index of the node in the textElement to split before
+ * @returns a Tupple of three BodyElements
+ */
+const splitTextElementAndInsertBetweenParts = (
+	textElement: Text,
+	elementToInsert: BodyElement,
+	splitPoint = 1,
+): [Text, BodyElement, Text] => {
+	// uses a clone of the original doc to avoid mutations
+	const childNodeArray = Array.from(
+		textElement.doc.cloneNode(true).childNodes,
+	);
+
+	const newDocs: [DocumentFragment, DocumentFragment] = [
+		JSDOM.fragment(''),
+		JSDOM.fragment(''),
+	];
+
+	childNodeArray.slice(0, Math.floor(splitPoint)).forEach((element) => {
+		newDocs[0].appendChild(element);
+	});
```

Comment:
I think we could use `append` like this:

```suggestion
	newDocs[0].append(...childNodeArray.slice(0, splitPoint))
```